lib/honeybadger/http_adapter.ex

```diff
@@ -0,0 +1,173 @@
+
+    defstruct http_adapter: nil, request_url: nil, status: 200, headers: [], body: ""
+
+    def format(response) do
+      [request_url | _rest] = String.split(response.request_url, "?", parts: 2)
+
+      """
+      HTTP Adapter: #{inspect(response.http_adapter)}
+      Request URL: #{request_url}
+
+      Response status: #{response.status}
+
+      Response headers:
+      #{Enum.reduce(response.headers, "", fn {k, v}, acc -> acc <> "\n#{k}: #{v}" end)}
+
+      Response body:
+      #{inspect(response.body)}
+      """
+    end
+  end
+
+  @type method :: :get | :post
+  @type body :: binary() | nil
+  @type headers :: [{binary(), binary()}]
+
+  @callback request(method(), binary(), body(), headers(), Keyword.t()) ::
+              {:ok, map()} | {:error, any()}
+
+  @doc """
+  Sets a user agent header.
+
+  The header value will be `Honeybadger-VERSION` with VERSION being the `:vsn` of
+  `:honeybadger` app.
+  """
+  @spec user_agent_header() :: {binary(), binary()}
+  def user_agent_header do
+    version = Application.spec(:honeybadger, :vsn) || "0.0.0"
+
+    {"User-Agent", "Honeybadger-#{version}"}
+  end
+
+  @default_http_client Enum.find_value(
+                         [
+                           {Req, Honeybadger.HTTPAdapter.Req},
+                           {Finch, Honeybadger.HTTPAdapter.Finch},
+                           {:hackney, Honeybadger.HTTPAdapter.Hackney}
+                         ],
+                         fn {dep, module} ->
+                           Code.ensure_loaded?(dep) && {module, []}
+                         end
+                       )
+
+  @doc """
+  Makes an HTTP request.
+
+  ## Options
+
+  - `:http_adapter` - The HTTP adapter to use, defaults to
+    `#{inspect(elem(@default_http_client, 0))}`
+  """
+  @spec request(atom(), binary(), binary() | nil, list(), Keyword.t()) ::
+          {:ok, HTTPResponse.t()} | {:error, HTTPResponse.t()} | {:error, term()}
+  def request(method, url, body, headers, opts) do
+    {http_adapter, http_adapter_opts} = get_adapter(opts)
+
+    method
+    |> http_adapter.request(url, body, headers, http_adapter_opts)
+    |> case do
+      {:ok, response} ->
+        decode_response(response, opts)
+
+      {:error, error} ->
+        {:error,
+         ServerUnreachableError.exception(
+           reason: error,
+           http_adapter: http_adapter,
+           request_url: url
+         )}
+    end
+    |> case do
+      {:ok, %{status: status} = resp} when status in 200..399 ->
+        {:ok, %{resp | http_adapter: http_adapter, request_url: url}}
+
+      {:ok, %{status: status} = resp} when status in 400..599 ->
+        {:error, %{resp | http_adapter: http_adapter, request_url: url}}
+
+      {:error, error} ->
+        {:error, error}
+    end
+  end
+
+  defp get_adapter(opts) do
+    default_http_adapter = Application.get_env(:honeybadger, :http_adapter, @default_http_client)
+
+    case Keyword.get(opts, :http_adapter, default_http_adapter) do
```

Comment:
The difference between `default_http_client` and `default_http_adapter` tripped me up briefly.

Reply:
Would @default_http_client_lib be a clearer name?